bee-crypto/src/ternary/sponge/batched_curl/bct.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+use std::slice::SliceIndex;
+
+#[derive(Clone)]
+pub struct BCTritBuf {
+    lo: Vec<usize>,
+    hi: Vec<usize>,
+}
+
+impl BCTritBuf {
+    pub fn as_slice(&self) -> BCTritRef<'_, [usize]> {
+        BCTritRef {
+            lo: &self.lo,
+            hi: &self.hi,
+        }
+    }
+
+    pub fn as_slice_mut(&mut self) -> BCTritMut<'_, [usize]> {
+        BCTritMut {
+            lo: &mut self.lo,
+            hi: &mut self.hi,
+        }
+    }
+
+    pub fn fill(&mut self, value: usize) {
+        for (lo, hi) in self.lo.iter_mut().zip(self.hi.iter_mut()) {
+            *lo = value;
+            *hi = value;
+        }
+    }
+
+    pub fn filled(value: usize, len: usize) -> Self {
+        BCTritBuf {
+            lo: vec![value; len],
+            hi: vec![value; len],
+        }
+    }
+
+    pub fn zeros(len: usize) -> Self {
+        Self::filled(0, len)
+    }
+
+    pub fn len(&self) -> usize {
+        self.lo.len()
+    }
+
+    pub fn lo(&self) -> &[usize] {
+        &self.lo
+    }
+
+    pub fn hi(&self) -> &[usize] {
+        &self.hi
+    }
+
+    pub fn lo_mut(&mut self) -> &mut [usize] {
+        &mut self.lo
+    }
+
+    pub fn hi_mut(&mut self) -> &mut [usize] {
+        &mut self.hi
+    }
+
+    pub fn get<I: SliceIndex<[usize]> + Clone>(&self, index: I) -> BCTritRef<I::Output> {
+        BCTritRef {
+            lo: &self.lo[index.clone()],
```

Comment:
Why is the clone necessary ?

Reply:
because `Index::index` receives `I` by value. With regular `usize` indices there should be no problem because those get copied. However, ranges aren't `Copy`so we have to clone them.

Reply:
Ahh forgot about ranges, thanks 👍